cirq-core/cirq/ops/raw_types.py

```diff
@@ -1112,3 +1078,78 @@ def _validate_qid_shape(val: Any, qubits: Sequence['cirq.Qid']) -> None:
         raise ValueError(
             f'Duplicate qids for <{val!r}>. Expected unique qids but got <{qubits!r}>.'
         )
+
+
+def _operations_commutes_impl(
+    ops1: Collection[Operation], ops2: Collection[Operation], *, atol: float
+) -> Union[bool, NotImplementedType]:
+    """Determine if two collections of non-overlapping Operations commute.
+
+    This function implements the commutes protocol for the Operation and Moment classes
+    and is not intended for other use.
+
+    Args:
+        ops1: The first collection of operations.  It is assumed each operation
+            acts on different qubits, i.e., the operations can form a Moment.
+        ops2: The second collection of operations to be checked for commutation
+            with `ops1`.  It is assumed each operation acts on different qubits,
+            i.e., the operations can form a Moment.
+        atol: Absolute error tolerance. If all entries in ops1@ops2 - ops2@ops1
+            have a magnitude less than this tolerance, ops1 and ops2 are considered
+            to commute.
+
+    Returns:
+        True: `ops1` and `ops2` commute (or approximately commute).
+        False: `ops1` and `ops2` do not commute.
+        NotImplemented: The commutativity cannot be determined here.
+    """
+    ops1_keys = frozenset(k for op in ops1 for k in protocols.measurement_key_objs(op))
+    ops2_keys = frozenset(k for op in ops2 for k in protocols.measurement_key_objs(op))
+    ops1_control_keys = frozenset(k for op in ops1 for k in protocols.control_keys(op))
+    ops2_control_keys = frozenset(k for op in ops2 for k in protocols.control_keys(op))
+    if (
+        not ops1_keys.isdisjoint(ops2_keys)
+        or not ops1_control_keys.isdisjoint(ops2_keys)
+        or not ops2_control_keys.isdisjoint(ops1_keys)
+    ):
+        return False
+
+    ops1_qubits = frozenset(q for op in ops1 for q in op.qubits)
```

Comment:
Kind of odd the disparity between qubits gotten via public member methods and cbits gotten from protocols. Seems like it should be consistent. (Preferably use members for speed). (Ultimately I think all those measurement key protocols should be phased out. They're not really what protocols were designed for. IIRC back when I looked into doing so, their use within the commutes protocol was a blocker for that. So if it's possible to avoid them here that would be nice).

Reply:
Ack.  `protocols.control_keys` calls on the `_control_keys_` method, which is available in [ClassicallyControlledOperation](https://github.com/quantumlib/Cirq/blob/a55b67516d6033028d360ca6c16c8bd1a822aa9e/cirq-core/cirq/ops/classically_controlled_operation.py#L239-L243), but I don't see a public method there or in the base class that would return the same.

On another page, it is faster to build the set of ops1_qubits with `set.union()` compared to a double-loop iteration so I am going to change that.